site/docs/plugins/conversations.md

```diff
@@ -7,1190 +7,1527 @@ next: false
-  );
+Also, just like with sessions, you can store your conversations data under a namespace using the `prefix` option.
+This is especially useful if you want to use the same storage adapter for both your session data and your conversations data.
+Storing the data in namespaces will prevent it from clashing.
 
-  // We now copy the responses to another chat for review.
-  await userHomeAddressContext.copyMessage(TEAM_REVIEW_CHAT);
-  await userNationalityContext.copyMessage(TEAM_REVIEW_CHAT);
-}
+You can specify both options as follows.
+
+```ts
+bot.use(conversations({
+  storage: {
+    type: "key",
+    adapter: storageAdapter,
+    getStorageKey: (ctx) => ctx.from?.id.toString(),
+    prefix: "convo-",
+  },
+}));
 ```
 
-:::
+If a conversation is entered for a user with user identifier `424242`, the storage key will now be `convo-424242`.
+
+Check out the API reference for [`ConversationStorage`](/ref/conversations/conversationstorage) to see more details about storing data with the conversations plugin.
+Among other things, it will explain how to store data without a storage key function at all using `type: "context"`.
+
+## Using Plugins Inside Conversations
 
-Usually, outside of the conversations plugin, each of these updates would be handled by the [middleware system](../guide/middleware) of your bot.
-Hence, your bot would handle the update via a context object which gets passed to your handlers.
+[Remember](#conversational-context-objects) that the context objects inside conversations are independent from the context objects in the surrounding middleware.
+This means that they will have no plugins installed on them by default---even if the plugins are installed on your bot.
 
-In conversations, you will obtain this new context object from the `wait` call.
-In turn, you can handle different updates differently based on this object.
-For example, you can check for text messages:
+Fortunately, all grammY plugins [except sessions](#accessing-sessions-inside-conversations) are compatible with conversations.
+For example, this is how you can install the [hydrate plugin](./hydrate) for a conversation.
 
 ::: code-group
 
 ```ts [TypeScript]
-async function waitForText(conversation: MyConversation, ctx: MyContext) {
-  // Wait for the next update:
-  ctx = await conversation.wait();
-  // Check for text:
-  if (ctx.message?.text) {
-    // ...
-  }
+// Only install the conversations plugin outside.
+type MyContext = ConversationFlavor<Context>;
+// Only install the hydrate plugin inside.
+type MyConversationContext = HydrateFlavor<Context>;
+
+bot.use(conversations());
+
+// Pass the outside and the inside context object.
+type MyConversation = Conversation<MyContext, MyConversationContext>;
+async function convo(conversation: MyConversation, ctx: MyConversationContext) {
+  // The hydrate plugin is installed on `ctx` here.
+  const other = await conversation.wait();
+  // The hydrate plugin is installed on `other` here, too.
 }
```

Comment:
The comment on this code doesn’t sound right. Could it be that a line of code is missing? Or maybe the comment is misplaced and should look like this:

```suggestion
// The hydrate plugin is installed on `ctx` here.
async function convo(conversation: MyConversation, ctx: MyConversationContext) {
  // The hydrate plugin is installed on `other` here, too.
  const other = await conversation.wait();
}
```

Reply:
It refers to the parameter `ctx` that is in scope here. Perhaps that should be clarified somehow, though.
```suggestion
async function convo(conversation: MyConversation, ctx: MyConversationContext) {
  // The hydrate plugin is installed on the parameter `ctx` here.
  const other = await conversation.wait();
  // The hydrate plugin is installed on `other` here, too.
}
```